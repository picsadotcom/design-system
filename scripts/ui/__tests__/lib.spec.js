// Copyright (c) 2015-present, salesforce.com, inc. All rights reserved
// Licensed under BSD 3-Clause - see LICENSE.txt or git.io/sfdc-license

/* eslint-env jest */

const {ui} = require("../");
const I = require("immutable");
const createInstance = require("../../lib.js");
const showcase = require("../../ui/showcase.js");
const React = require("react");
const ReactDOM = require("react-dom/server");

describe("scripts/lib.js", () => {
  let uiJson, SLDS;

  beforeAll(() => {
    ui().fork(
      e => {
        throw e;
      },
      r => {
        uiJson = r
        SLDS = createInstance(uiJson)
      }
    );
  });

  it("returns a list of components", () => {
    const comps = SLDS.components()
    expect(comps.count()).toBeGreaterThan(1)
    expect(comps.first()).toEqual('buttons')
  });

  it("returns a component", () => {
    const comp = SLDS.component('buttons').getOrElse(null)
    expect(comp.get('id')).toEqual('buttons')
    expect(comp.get('restrictees').count()).toBeGreaterThan(0)
  });

  it("finds a variant", () => {
    const comp = SLDS.component('buttons').getOrElse(null)
    const variant = SLDS.findVariant(comp, 'stateful').getOrElse(null)
    expect(variant.get('id')).toEqual('stateful')
    expect(variant.get('restrictees').count()).toBeGreaterThan(0)
  });

  it("gets every example in the system", () => {
    const allMarkup =
    uiJson.flatMap((group, name) =>
      group.map(item =>
        SLDS.variants(item)
        .flatMap(variant =>
          showcase(item.get('id'), variant.get('id'), name === 'utilities', true)
          .getOrElse(I.List())
          .flatMap(section =>
            section.get('items')
            .map(i =>
              React.isValidElement(i.get('element'))
              ? ReactDOM.renderToStaticMarkup(i.get('element'))
              : `FAILED: ${item.get('id')}/${variant.get('id')}/${i.get('id')}`
            )
          )
        )
      )
    )
    expect(allMarkup).toMatchSnapshot();
  })
});
