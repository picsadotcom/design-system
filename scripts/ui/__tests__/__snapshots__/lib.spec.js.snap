// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scripts/lib.js gets every example in the system 1`] = `
Object {
  "alignment": Array [
    "<div class=\\"slds-align_absolute-center\\">This content will be positioned in the absolute center of its container</div>",
  ],
  "buttons": Array [
    "FAILED: buttons/base/states",
    "FAILED: buttons/with-icon/examples",
  ],
  "data-tables": Array [
    "<table class=\\"slds-table slds-table_bordered slds-table_cell-buffer slds-table_striped\\"><thead><tr class=\\"slds-text-title_caps\\"><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Opportunity Name\\">Opportunity Name</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Account Name\\">Account Name</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Close Date\\">Close Date</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Stage\\">Stage</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Confidence\\">Confidence</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Amount\\">Amount</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Contact\\">Contact</div></th></tr></thead><tbody><tr><th scope=\\"row\\" data-label=\\"Opportunity Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub\\"><a href=\\"javascript:void(0);\\">Cloudhub</a></div></th><td data-label=\\"Account Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub\\">Cloudhub</div></td><td data-label=\\"Close Date\\"><div class=\\"slds-truncate\\" title=\\"4/14/2015\\">4/14/2015</div></td><td data-label=\\"Prospecting\\"><div class=\\"slds-truncate\\" title=\\"Prospecting\\">Prospecting</div></td><td data-label=\\"Confidence\\"><div class=\\"slds-truncate\\" title=\\"20%\\">20%</div></td><td data-label=\\"Amount\\"><div class=\\"slds-truncate\\" title=\\"$25k\\">$25k</div></td><td data-label=\\"Contact\\"><div class=\\"slds-truncate\\" title=\\"jrogers@cloudhub.com\\"><a href=\\"javascript:void(0);\\">jrogers@cloudhub.com</a></div></td></tr><tr><th scope=\\"row\\" data-label=\\"Opportunity Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub + Anypoint Connectors\\"><a href=\\"javascript:void(0);\\">Cloudhub + Anypoint Connectors</a></div></th><td data-label=\\"Account Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub\\">Cloudhub</div></td><td data-label=\\"Close Date\\"><div class=\\"slds-truncate\\" title=\\"4/14/2015\\">4/14/2015</div></td><td data-label=\\"Prospecting\\"><div class=\\"slds-truncate\\" title=\\"Prospecting\\">Prospecting</div></td><td data-label=\\"Confidence\\"><div class=\\"slds-truncate\\" title=\\"20%\\">20%</div></td><td data-label=\\"Amount\\"><div class=\\"slds-truncate\\" title=\\"$25k\\">$25k</div></td><td data-label=\\"Contact\\"><div class=\\"slds-truncate\\" title=\\"jrogers@cloudhub.com\\"><a href=\\"javascript:void(0);\\">jrogers@cloudhub.com</a></div></td></tr><tr><th scope=\\"row\\" data-label=\\"Opportunity Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub\\"><a href=\\"javascript:void(0);\\">Cloudhub</a></div></th><td data-label=\\"Account Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub\\">Cloudhub</div></td><td data-label=\\"Close Date\\"><div class=\\"slds-truncate\\" title=\\"4/14/2015\\">4/14/2015</div></td><td data-label=\\"Prospecting\\"><div class=\\"slds-truncate\\" title=\\"Prospecting\\">Prospecting</div></td><td data-label=\\"Confidence\\"><div class=\\"slds-truncate\\" title=\\"20%\\">20%</div></td><td data-label=\\"Amount\\"><div class=\\"slds-truncate\\" title=\\"$25k\\">$25k</div></td><td data-label=\\"Contact\\"><div class=\\"slds-truncate\\" title=\\"jrogers@cloudhub.com\\"><a href=\\"javascript:void(0);\\">jrogers@cloudhub.com</a></div></td></tr></tbody></table>",
    "<table class=\\"slds-table slds-table_bordered slds-table_cell-buffer slds-no-row-hover\\"><thead><tr class=\\"slds-text-title_caps\\"><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Opportunity Name\\">Opportunity Name</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Account Name\\">Account Name</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Close Date\\">Close Date</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Stage\\">Stage</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Confidence\\">Confidence</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Amount\\">Amount</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Contact\\">Contact</div></th></tr></thead><tbody><tr><th scope=\\"row\\" data-label=\\"Opportunity Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub\\"><a href=\\"javascript:void(0);\\">Cloudhub</a></div></th><td data-label=\\"Account Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub\\">Cloudhub</div></td><td data-label=\\"Close Date\\"><div class=\\"slds-truncate\\" title=\\"4/14/2015\\">4/14/2015</div></td><td data-label=\\"Prospecting\\"><div class=\\"slds-truncate\\" title=\\"Prospecting\\">Prospecting</div></td><td data-label=\\"Confidence\\"><div class=\\"slds-truncate\\" title=\\"20%\\">20%</div></td><td data-label=\\"Amount\\"><div class=\\"slds-truncate\\" title=\\"$25k\\">$25k</div></td><td data-label=\\"Contact\\"><div class=\\"slds-truncate\\" title=\\"jrogers@cloudhub.com\\"><a href=\\"javascript:void(0);\\">jrogers@cloudhub.com</a></div></td></tr><tr><th scope=\\"row\\" data-label=\\"Opportunity Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub + Anypoint Connectors\\"><a href=\\"javascript:void(0);\\">Cloudhub + Anypoint Connectors</a></div></th><td data-label=\\"Account Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub\\">Cloudhub</div></td><td data-label=\\"Close Date\\"><div class=\\"slds-truncate\\" title=\\"4/14/2015\\">4/14/2015</div></td><td data-label=\\"Prospecting\\"><div class=\\"slds-truncate\\" title=\\"Prospecting\\">Prospecting</div></td><td data-label=\\"Confidence\\"><div class=\\"slds-truncate\\" title=\\"20%\\">20%</div></td><td data-label=\\"Amount\\"><div class=\\"slds-truncate\\" title=\\"$25k\\">$25k</div></td><td data-label=\\"Contact\\"><div class=\\"slds-truncate\\" title=\\"jrogers@cloudhub.com\\"><a href=\\"javascript:void(0);\\">jrogers@cloudhub.com</a></div></td></tr></tbody></table>",
    "<table class=\\"slds-table slds-table_bordered slds-table_cell-buffer slds-table_col-bordered\\"><thead><tr class=\\"slds-text-title_caps\\"><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Opportunity Name\\">Opportunity Name</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Account Name\\">Account Name</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Close Date\\">Close Date</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Stage\\">Stage</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Confidence\\">Confidence</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Amount\\">Amount</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Contact\\">Contact</div></th></tr></thead><tbody><tr><th scope=\\"row\\" data-label=\\"Opportunity Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub\\"><a href=\\"javascript:void(0);\\">Cloudhub</a></div></th><td data-label=\\"Account Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub\\">Cloudhub</div></td><td data-label=\\"Close Date\\"><div class=\\"slds-truncate\\" title=\\"4/14/2015\\">4/14/2015</div></td><td data-label=\\"Prospecting\\"><div class=\\"slds-truncate\\" title=\\"Prospecting\\">Prospecting</div></td><td data-label=\\"Confidence\\"><div class=\\"slds-truncate\\" title=\\"20%\\">20%</div></td><td data-label=\\"Amount\\"><div class=\\"slds-truncate\\" title=\\"$25k\\">$25k</div></td><td data-label=\\"Contact\\"><div class=\\"slds-truncate\\" title=\\"jrogers@cloudhub.com\\"><a href=\\"javascript:void(0);\\">jrogers@cloudhub.com</a></div></td></tr><tr><th scope=\\"row\\" data-label=\\"Opportunity Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub + Anypoint Connectors\\"><a href=\\"javascript:void(0);\\">Cloudhub + Anypoint Connectors</a></div></th><td data-label=\\"Account Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub\\">Cloudhub</div></td><td data-label=\\"Close Date\\"><div class=\\"slds-truncate\\" title=\\"4/14/2015\\">4/14/2015</div></td><td data-label=\\"Prospecting\\"><div class=\\"slds-truncate\\" title=\\"Prospecting\\">Prospecting</div></td><td data-label=\\"Confidence\\"><div class=\\"slds-truncate\\" title=\\"20%\\">20%</div></td><td data-label=\\"Amount\\"><div class=\\"slds-truncate\\" title=\\"$25k\\">$25k</div></td><td data-label=\\"Contact\\"><div class=\\"slds-truncate\\" title=\\"jrogers@cloudhub.com\\"><a href=\\"javascript:void(0);\\">jrogers@cloudhub.com</a></div></td></tr></tbody></table>",
    "<table class=\\"slds-table slds-table_bordered slds-table_cell-buffer\\"><thead><tr class=\\"slds-text-title_caps\\"><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Opportunity Name\\">Opportunity Name</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Account Name\\">Account Name</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Close Date\\">Close Date</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Stage\\">Stage</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Confidence\\">Confidence</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Amount\\">Amount</div></th><th scope=\\"col\\"><div class=\\"slds-truncate\\" title=\\"Contact\\">Contact</div></th></tr></thead><tbody><tr><th scope=\\"row\\" data-label=\\"Opportunity Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub\\"><a href=\\"javascript:void(0);\\">Cloudhub</a></div></th><td data-label=\\"Account Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub\\">Cloudhub</div></td><td data-label=\\"Close Date\\"><div class=\\"slds-truncate\\" title=\\"4/14/2015\\">4/14/2015</div></td><td data-label=\\"Prospecting\\"><div class=\\"slds-truncate\\" title=\\"Prospecting\\">Prospecting</div></td><td data-label=\\"Confidence\\"><div class=\\"slds-truncate\\" title=\\"20%\\">20%</div></td><td data-label=\\"Amount\\"><div class=\\"slds-truncate\\" title=\\"$25k\\">$25k</div></td><td data-label=\\"Contact\\"><div class=\\"slds-truncate\\" title=\\"jrogers@cloudhub.com\\"><a href=\\"javascript:void(0);\\">jrogers@cloudhub.com</a></div></td></tr><tr><th scope=\\"row\\" data-label=\\"Opportunity Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub + Anypoint Connectors\\"><a href=\\"javascript:void(0);\\">Cloudhub + Anypoint Connectors</a></div></th><td data-label=\\"Account Name\\"><div class=\\"slds-truncate\\" title=\\"Cloudhub\\">Cloudhub</div></td><td data-label=\\"Close Date\\"><div class=\\"slds-truncate\\" title=\\"4/14/2015\\">4/14/2015</div></td><td data-label=\\"Prospecting\\"><div class=\\"slds-truncate\\" title=\\"Prospecting\\">Prospecting</div></td><td data-label=\\"Confidence\\"><div class=\\"slds-truncate\\" title=\\"20%\\">20%</div></td><td data-label=\\"Amount\\"><div class=\\"slds-truncate\\" title=\\"$25k\\">$25k</div></td><td data-label=\\"Contact\\"><div class=\\"slds-truncate\\" title=\\"jrogers@cloudhub.com\\"><a href=\\"javascript:void(0);\\">jrogers@cloudhub.com</a></div></td></tr></tbody></table>",
    "FAILED: data-tables/advanced/states",
    "FAILED: data-tables/inline-edit/states",
  ],
  "floats": Array [
    "<div class=\\"slds-clearfix\\"><div class=\\"slds-float_left\\"><p>I’m floooaaaating</p></div></div>",
    "<div class=\\"slds-clearfix\\"><div class=\\"slds-float_right\\"><p>I’m floooaaaating</p></div></div>",
    "<div class=\\"slds-float_none\\"><p>I’m not floooaaaating</p></div>",
    "<div class=\\"slds-clearfix\\"><div class=\\"slds-float_left\\"><p>I’m floooaaaating</p></div><div class=\\"slds-float_right\\"><p>I’m floooaaaating</p></div></div>",
    "<div><p>I’m floooaaaating</p></div>",
    "<div class=\\"slds-clearfix\\"><div class=\\"slds-float_left\\"><p>I’m floooaaaating</p></div></div>",
    "<div class=\\"slds-clearfix\\"><div class=\\"slds-float_right\\"><p>I’m floooaaaating</p></div></div>",
    "<div class=\\"slds-float_none\\"><p>I’m not floooaaaating</p></div>",
    "<div class=\\"slds-clearfix\\"><div class=\\"slds-float_left\\"><p>I’m floooaaaating</p></div><div class=\\"slds-float_right\\"><p>I’m floooaaaating</p></div></div>",
    "<div><p>I’m floooaaaating</p></div>",
    "<div class=\\"slds-clearfix\\"><div class=\\"slds-float_left\\"><p>I’m floooaaaating</p></div></div>",
    "<div class=\\"slds-clearfix\\"><div class=\\"slds-float_right\\"><p>I’m floooaaaating</p></div></div>",
    "<div class=\\"slds-float_none\\"><p>I’m not floooaaaating</p></div>",
    "<div class=\\"slds-clearfix\\"><div class=\\"slds-float_left\\"><p>I’m floooaaaating</p></div><div class=\\"slds-float_right\\"><p>I’m floooaaaating</p></div></div>",
    "<div><p>I’m floooaaaating</p></div>",
  ],
}
`;
